internal/dnsdocs/cli_help/generator.go

```diff
@@ -0,0 +1,96 @@
+package main
+
+//go:generate go run .
+
+import (
+	"bytes"
+	"log"
+	"os"
+	"strings"
+	"text/template"
+
+	"github.com/go-acme/lego/v4/cmd"
+	"github.com/urfave/cli/v2"
+)
+
+const outputFile = "../../../docs/data/zz_cli_help.toml"
+
+var outputTpl = template.Must(template.New("output").Parse(`# THIS FILE IS AUTO-GENERATED. PLEASE DO NOT EDIT.
+
+{{ range .}}
+[[command]]
+title   = "{{.Title}}"
+content = """
+{{.Help}}
+"""
+{{end -}}
+`))
+
+type commandHelp struct{ Title, Help string }
+
+func main() {
+	log.SetFlags(0)
+
+	// collect output of various help pages
+	var help []commandHelp
+	for _, args := range [][]string{
+		{"lego", "help"},
+		{"lego", "help", "run"},
+		{"lego", "help", "renew"},
+		{"lego", "help", "revoke"},
+		{"lego", "help", "list"},
+		{"lego", "dnshelp"},
+	} {
+		content, err := run(args)
+		if err != nil {
+			log.Fatalf("running %s failed: %v", args, err)
+		}
+		help = append(help, content)
+	}
+
+	f, err := os.OpenFile(outputFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o644)
+	if err != nil {
+		log.Fatalf("cannot open cli_help.toml: %v", err)
+	}
+	if err = outputTpl.Execute(f, help); err != nil {
+		f.Close()
+		log.Fatalf("failed to write cli_help.toml: %v", err)
+	}
+	f.Close()
+
+	log.Println("cli_help.toml updated")
+}
+
+var lego = func() *cli.App {
+	// Construct cli app, very similar to cmd/lego/main.go.
+	// Notable differences:
+	// - substitute "." for CWD in default config path, as the user
+	//   will very likely see a different path
+	// - do not include version information, because we're likely
+	//   running against a snapshot
+	// - skip DNS help and provider list, as initialization takes time,
+	//   and we don't generate `lego dns --help` here
+	app := cli.NewApp()
+	app.Name = "lego"
+	app.HelpName = "lego"
+	app.Usage = "Let's Encrypt client written in Go"
+	app.Flags = cmd.CreateFlags("./.lego")
+	app.Commands = cmd.CreateCommands()
+	return app
+}()
```

Comment:
:thinking: Maybe this code should live in `cmd/`, and reused by `cmd/lego/main.go`? Something like

<details><summary>cmd/cmd_app.go</summary>

```go
func InitApp(path string) *cli.App {
	app := cli.NewApp()
	app.Name = "lego"
	app.HelpName = "lego"
	app.Usage = "Let's Encrypt client written in Go"
	app.Flags = CreateFlags(path)
	app.Commands = CreateCommands()
	return app
}
```

</details>
<details><summary>cmd/lego/main.go</summary>

```go
func main() {
	var defaultPath string
	cwd, err := os.Getwd()
	if err == nil {
		defaultPath = filepath.Join(cwd, ".lego")
	}

	app := cmd.InitApp(defaultPath)

	app.Version = version
	cli.VersionPrinter = func(c *cli.Context) {
		fmt.Printf("lego version %s %s/%s\n", c.App.Version, runtime.GOOS, runtime.GOARCH)
	}
	app.EnableBashCompletion = true
	app.Before = cmd.Before

	err = app.Run(os.Args)
	if err != nil {
		log.Fatal(err)
	}
}
```

</details>
<details><summary>internal/dnsdocs/cli_help/generator.go</summary>

```go
var lego = cmd.InitApp("./.lego")
```

</details>

OTOH, how likely is it that the CLI app initialization changes so much, that the documentation becomes unusable? The most important part (printing the flags) would be very much unaffected by most changes to the `*cli.App`.

Reply:
I think, I'll leave it as is. "[A little copying is better than a little dependency.](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=9m28s)"